index.js

```diff
@@ -528,80 +574,140 @@ Client.prototype._destroy = function (err, cb) {
+  // - It `gzipStream.end()`s the stream when:
+  //   (a) approximately `apiRequestSize` of data have been written,
+  //   (b) `apiRequestTime` seconds have passed, or
+  //   (c) `_chopper.chop()` is explicitly called via `client.flush()`,
+  //       e.g. used by the Node.js APM agent after `client.sendError()`.
+  // - This function makes the HTTP POST to the apm-server, pipes the gzipStream
+  //   to it, and waits for the completion of the request and the apm-server
+  //   response.
+  // - Then it calls the given `next` callback to signal StreamChopper that
+  //   another chopped stream can be created, when there is more the send.
+  //
+  // Of course, things can go wrong. Here are the known ways this pipeline can
+  // conclude.
+  // - intake response success - A successful response from the APM server. This
+  //   is the normal operation case described above.
+  // - gzipStream error - An "error" event on the gzip stream.
+  // - intake request error - An "error" event on the intake HTTP request, e.g.
+  //   ECONNREFUSED or ECONNRESET.
+  // - intakeResTimeout - A timer started *after* we are finished sending data
+  //   to the APM server by which we require a response (including its body). By
+  //   default this is 10s -- a very long time to allow for a slow or far
+  //   apm-server. If we hit this, APM server is problematic anyway, so the
+  //   delay doesn't add to the problems.
+  // - serverTimeout - An idle timeout value (default 30s) set on the socket.
+  //   This is a catch-all fallback for an otherwised wedged connection. If this
+  //   is being hit, there is some major issue in the application (possibly a
+  //   bug in the APM agent).
+  // - process completion - The Client takes pains to always `.unref()` its
+  //   handles to never keep a using process open if it is ready to exit. When
+  //   the process is ready to exit, the following happens:
+  //    - The "beforeExit" handler above will call `client.end()`,
+  //    - which calls `client._ref()` (to *hold the process open* to complete
+  //      this request), then `_chopper.end()` to end the `gzipStream` so
+  //      this request can complete soon.
+  //    - We then expect this request to complete quickly and the process will
+  //      then finish exiting. A subtlety is if the APM server is not responding
+  //      then we'll wait on `intakeResTimeoutOnEnd` (by default 1s).
+  return function makeIntakeRequest (gzipStream, next) {
+    const reqId = crypto.randomBytes(16).toString('hex')
+    const log = client._log.child({ reqId })
+    const startTime = process.hrtime()
+    const timeline = []
+    let bytesWritten = 0
+    let intakeRes
+    let intakeResTimer = null
+    const intakeResTimeout = client._conf.intakeResTimeout
+    const intakeResTimeoutOnEnd = client._conf.intakeResTimeoutOnEnd
+
+    // `_active` is used to coordinate the callback to `client.flush(db)`.
+    client._active = true
```

Comment:
Yup, the code is designed to only have one intake request active at a time. Otherwise there is a bug. The `completePart()` function is written to handle all the (documented in its top comment) ways the 3 streams can conclude, and only call StreamChopper's `next()` when those 3 streams are all cleanly finished or destroyed.

Note that this `client._active` handling was in the earlier `onStream` code. However, because the earlier code didn't wait on the completion of the HTTP response from APM server it *was* possible to have multiple intake requests going at the same time.